File: src/modules/trees/traversals/1_recursive_dfs.ts
class TreeNode {
    left: TreeNode | null
    right: TreeNode | null
    data: number
    constructor(data: number) {
        this.data = data
        this.left = null
        this.right = null
    }
}

function inorderTraversal(root:TreeNode|null):void{
    if(root === null) return;
    inorderTraversal(root.left)
    console.log(root.data)
    inorderTraversal(root.right)
}

function preorderTraversal(root:TreeNode|null):void{
    if(root === null) return;
    console.log(root.data)
    preorderTraversal(root.left)
    preorderTraversal(root.right)
}

function postorderTraversal(root:TreeNode|null):void{
    if(root === null) return;
    postorderTraversal(root.left)
    postorderTraversal(root.right)
    console.log(root.data)
}

let root = new TreeNode(10);
root.left = new TreeNode(5);
root.right = new TreeNode(20);
root.left.left = new TreeNode(3);
root.left.right = new TreeNode(8);
root.right.left = new TreeNode(15);
root.right.right = new TreeNode(25);

console.log("Inorder Traversal: ");
inorderTraversal(root);
console.log();

console.log("Preorder Traversal: ");
preorderTraversal(root);
console.log();

console.log("Postorder Traversal: ");
postorderTraversal(root);

export {}

File: src/modules/trees/2_traversals/1_recursive_dfs.go
package main

import "fmt"

type TreeNode struct {
	left  *TreeNode
	right *TreeNode
	data  int
}

type DfsTraversal interface {
	inorderTraversal(root *TreeNode)
	preorderTraversal(root *TreeNode)
	postorderTraversal(root *TreeNode)
}

func NewTreeNode(data int) *TreeNode {
	return &TreeNode{data: data}
}

type Tree struct{}

func (t *Tree) inorderTraversal(root *TreeNode) {
	if root == nil {
		return
	}

	t.inorderTraversal(root.left)
	fmt.Println(root.data)
	t.inorderTraversal(root.right)
}

func (t *Tree) preorderTraversal(root *TreeNode) {
	if root == nil {
		return
	}

	fmt.Println(root.data)
	t.preorderTraversal(root.left)
	t.preorderTraversal(root.right)
}

func (t *Tree) postorderTraversal(root *TreeNode) {
	if root == nil {
		return
	}

	t.postorderTraversal(root.left)
	t.postorderTraversal(root.right)
	fmt.Println(root.data)
}

func main() {
	// Create nodes
	root := NewTreeNode(1)
	root.left = NewTreeNode(2)
	root.right = NewTreeNode(3)
	root.left.left = NewTreeNode(4)
	root.left.right = NewTreeNode(5)
	root.right.left = NewTreeNode(6)
	root.right.right = NewTreeNode(7)

	// Initialize the tree
	var tree DfsTraversal = &Tree{}

	// Perform different DFS traversals
	fmt.Println("Inorder Traversal:")
	tree.inorderTraversal(root)

	fmt.Println("Preorder Traversal:")
	tree.preorderTraversal(root)

	fmt.Println("Postorder Traversal:")
	tree.postorderTraversal(root)
}
